cannon/serialize/binary.go

```diff
@@ -0,0 +1,75 @@
+package serialize
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+
+	"github.com/ethereum-optimism/optimism/op-service/ioutil"
+)
+
+// Serializable defines functionality for a type that may be serialized to raw bytes.
+type Serializable interface {
+	// Serialize encodes the type as raw bytes.
+	Serialize(out io.Writer) error
+
+	// Deserialize decodes raw bytes into the type.
+	Deserialize(in io.Reader) error
+}
+
+func LoadSerializedBinary[X any](inputPath string) (*X, error) {
+	if inputPath == "" {
+		return nil, errors.New("no path specified")
+	}
+	var f io.ReadCloser
+	f, err := ioutil.OpenDecompressed(inputPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open file %q: %w", inputPath, err)
+	}
+	defer f.Close()
+
+	var x X
+	serializable, ok := reflect.ValueOf(&x).Interface().(Serializable)
+	if !ok {
+		return nil, fmt.Errorf("%T is not a Serializable", x)
+	}
+	err = serializable.Deserialize(f)
+	if err != nil {
+		return nil, err
+	}
+	return &x, nil
+}
+
+func WriteSerializedBinary(outputPath string, value Serializable, perm os.FileMode) error {
+	if outputPath == "" {
+		return nil
+	}
+	var out io.Writer
+	finish := func() error { return nil }
+	if outputPath != "-" {
```

Comment:
(nit) maybe clearer to switch the if/else and assert on the positive match:
```suggestion
	if outputPath == "-" { ... }
```